### PR TITLE
Balance over pools using AWS ASGs

### LINKED ISSUE #3415: Balance across spot pools in deployed Kubernetes ASGs
┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-777)
┆Epic: Scalable kubernetes 
┆Issue Number: TOIL-777

### PR DESCRIPTION
**Pull request by adamnovak**
Fixes #3415, at least for balancing across node types. Balancing across regions is still TODO.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * AWS managed autoscaling can now balance across instance types when using `/`-separated lists of equivalent instance types (e.g. `--nodeTypes m5.8xlarge/m5a.8xlarge:0.50,m4.8xlarge`)

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/cliOptions.rst`
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

**Comment by DailyDreaming** [comment]
@adamnovak Is this ready for review?

**Comment by adamnovak** [comment]
@DailyDreaming Yes, this is ready for review.

**Comment by adamnovak** [comment]
No, sorry, this is not ready for review because actually it is broken.

**Comment by adamnovak** [comment]
OK, I've rebased this over to master without #3482 in it, and fixed the places where the Travis tests were reaching into the provisioner in ways that wouldn't work anymore. *Now* it should be ready for review.

**Comment by DailyDreaming** [comment]
@adamnovak I'll give a review sometime tomorrow.

File: src/toil/test/provisioners/aws/awsProvisionerTest.py
Comment (DailyDreaming) [inline src/toil/test/provisioners/aws/awsProvisionerTest.py:378]: Is this supposed to be semicolon separated?

**Review by DailyDreaming — COMMENTED** [review]
This looks good so far, and I'm excited to have this in.  Left some comments, though most are nits.  The mixing of the terms "node types" and "instance types" is unfortunate in the code, though I can't think of a better way to reference them, so I think we're stuck with it.  Thanks for some of the renaming, which I think helps quite a bit!

File: docs/running/cliOptions.rst
Comment (adamnovak) [reply docs/running/cliOptions.rst:151]: I don't think I quite meant to distinguish between node and instance types like that, but that distinction definitely makes sense.

I'm not sure what kind of language we could put in a command line argument for specifying these equivalent sets that is both concise and resistant to typos. I guess we could group equivalent instance types in parentheses or braces, but then people have to type that and we have to parse it.

Really what we would want is for Toil to already know which instance types are equivalent, based on its knowledge of instance shapes, and to automatically add in instance types it thinks are equivalent if you let it. But that might want to be its own ticket.

I'll rewrite the help text and synchronize the two --nodeTypes options.

File: src/toil/lib/ec2.py
Comment (adamnovak) [reply src/toil/lib/ec2.py:525]: The only way to hook multiple instance types onto one autoscaling group is via overrides. And we're using additional templates so that node storage overrides can work. I will note that.

File: src/toil/provisioners/__init__.py
Comment (adamnovak) [reply src/toil/provisioners/__init__.py]: Good idea

File: src/toil/provisioners/__init__.py
Comment (adamnovak) [reply src/toil/provisioners/__init__.py:85]: Sure.

File: src/toil/provisioners/__init__.py
Comment (adamnovak) [reply src/toil/provisioners/__init__.py]: Sure, I can do that.

File: src/toil/provisioners/aws/awsProvisioner.py
Comment (adamnovak) [reply src/toil/provisioners/aws/awsProvisioner.py:329]: I've renamed it to some more appropriate names.

File: src/toil/provisioners/aws/awsProvisioner.py
Comment (adamnovak) [reply src/toil/provisioners/aws/awsProvisioner.py:470]: I coalesced this duplicated code, although with documentation it came out slightly longer overall.

File: src/toil/provisioners/aws/awsProvisioner.py
Comment (adamnovak) [reply src/toil/provisioners/aws/awsProvisioner.py:473]: This chooses an arbitrary instance type from the set the user said were equivalent for them. I think it's up to how Python implements `set` whether it's always going to be the same one.

Since we aren't using an AWS ASG here, there's no way to actually balance across the spot pools. Since the set of instance types are supposed to be all the same to the user, we can just pick one arbitrarily. If that creates problems, we could move to a real random choice and spray across all the options.

We *could* write balancing code for Toil-workflow-managed Mesos autoscaling, but since I think it is going away I don't think it would be worth the effort. We want to use the new Kubernetes-based scaling specifically because it is smart enough to do stuff like that for us.

File: src/toil/provisioners/clusterScaler.py
Comment (adamnovak) [reply src/toil/provisioners/clusterScaler.py:483]: I think the way it is makes more sense. I guess the other way distinguishes better between the node type and instance type concepts. But if we made it say "Instances of type blah" that would be weird, since Toil talks about nodes throughout. And saying that the nodes have "instance_type blah" is kind of ugly.

File: src/toil/test/provisioners/aws/awsProvisionerTest.py
Comment (adamnovak) [reply src/toil/test/provisioners/aws/awsProvisionerTest.py:378]: Nope!

**Comment by adamnovak** [comment]
Hey @DailyDreaming, what do you think of my changes to address your comments?

**Comment by adamnovak** [comment]
A test of Mesos-cluster spot autoscaling failed because the spot instance never had its appliance come up in time. I suspect this test might be flaky? I will see if it happens again.

**Comment by DailyDreaming** [comment]
@adamnovak Will get a chance to review this tomorrow.

**Comment by adamnovak** [comment]
Sounds good. I'll update the PR, even though I think it is going to fail the Cactus test since that isn't fixed yet.

**Comment by adamnovak** [comment]
This is now failing because CWL 1.2 test 279 managed to fail, on single machine in https://ucsc-ci.com/databiosphere/toil/-/jobs/83594

I don't think that has anything to do with the provisioner changes here, so I am going to rerun it, but I am also going to see if I can break it in isolation.

**Comment by adamnovak** [comment]
I couldn't get CWL 1.2 test 279 to fail in 43 iterations on Courtyard. I think it has a very rare intermittent failure unrelated to this PR. Maybe some other CWL tests that fail more often get at the problem better?

Anyway, I think once you review this @DailyDreaming we can merge it. It's currently behind by some MyPy changes but we can just see if the merged MyPy passes and not wait for the rest of the merged test suite.

**Comment by adamnovak** [comment]
@DailyDreaming Did you want to make more comments on this, or should I just merge it?

**Review by DailyDreaming — APPROVED** [review]
Nope.  This looks good to me!

**Pull request merged**
